Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -1326,10 +1338,75 @@ extension SwiftLanguageServer {
       }
     }
   }
+
+  // Must be called on self.queue
+  public func _documentDiagnostic(
+    _ uri: DocumentURI,
+    _ completion: @escaping (Result<[Diagnostic], ResponseError>) -> Void
+  ) {
+    dispatchPrecondition(condition: .onQueue(queue))
+
+    guard let snapshot = documentManager.latestSnapshot(uri) else {
+      let msg = "failed to find snapshot for url \(uri)"
+      log(msg)
+      return completion(.failure(.unknown(msg)))
+    }
+
+    let keys = self.keys
+
+    let skreq = SKDRequestDictionary(sourcekitd: self.sourcekitd)
+    skreq[keys.request] = requests.diagnostics
+    skreq[keys.sourcefile] = snapshot.document.uri.pseudoPath
+
+    // FIXME: SourceKit should probably cache this for us.
+    if let compileCommand = self.commandsByFile[uri] {
+      skreq[keys.compilerargs] = compileCommand.compilerArgs
+    }
+
+    let supportsCodeDescription = capabilityRegistry.clientHasDiagnosticsCodeDescriptionSupport
+
+    let handle = self.sourcekitd.send(skreq, self.queue) { response in
+      guard let dict = response.success else {
+        return completion(.failure(ResponseError(response.failure!)))
+      }
+
+      var diagnostics: [Diagnostic] = []
+      dict[keys.diagnostics]?.forEach { _, diag in
+        if let diagnostic = Diagnostic(diag, in: snapshot, useEducationalNoteAsCode: supportsCodeDescription) {
+          diagnostics.append(diagnostic)
+        }
+        return true
+      }
+
+      completion(.success(diagnostics))
+    }
+
+    // FIXME: cancellation
+    _ = handle
```

Comment:
I think you should be able to implement cancellation similar to here

https://github.com/apple/sourcekit-lsp/blob/b7d427069c869efa1ae8764879935e6b877dab59/Sources/SourceKitLSP/Swift/OpenInterface.swift#L97-L101

Reply:
Could we leave this to a PR that fixes all cancellation at once? This was copied from [many places](https://github.com/search?q=repo%3Aapple%2Fsourcekit-lsp%20%2F%2F%20FIXME%3A%20cancellation&type=code)

Reply:
Sounds good to me as well